feat: show issue/PR author-role indicators in UI, after author logins

sideshowbarker opened the pull request:
# Summary

This change causes author-role indicators (colored Nerd Font icons) to be shown the dashboard, after the author’s GitHub login. Each indicator shows the author’s role relative to the repo to which they submitted their PR or issue. Fixes https://github.com/dlvhdr/gh-dash/issues/515.

For example, for a PR that’s the first PR the author has contributed to a particular repo, an indicator that the author is a first-time contributor to that repo will be shown.

Other indicators are used for showing that the author has previously contributed to the repo for the PR or issue, or that the author is member or owner of that repo, or a collaborator for that repo.

See https://docs.github.com/en/graphql/reference/enums#commentauthorassociation for details about the possible role (“author association”) values, or see the writeup at https://michaelheap.com/github-actions-check-permission/.

> [!Note]
> The screenshots below show how the author-role-indicator Nerd Font icons would look in the UI after this change.
>
> I don’t feel strongly at all about gh-dash needing to use the particular set of Nerd Font icons I chose in the current patch in this PR branch. So, @dlvhdr: If you have a different set of icons that you’d prefer — to best match with the rest of the existing gh-dash UI design — (or other commenters here have some good alternative suggestions), then I’m 100% fine with some other alternatives might be decided on instead.

## How did you test this change?

I ran `make test` (`go test -v ./...`), to ensure this change didn’t regress any existing tests.

Otherwise, I just tested this change manually by building this PR branch and then switching my config between the compact UI view and the non-compact UI view, and visually inspecting the displayed data to confirm that the author-role indicators are displayed as expected.

## Images/Videos

![image](https://github.com/user-attachments/assets/dcbb46fd-4439-4e43-912c-12751fbfaf6b)

---

<img width="922" alt="image" src="https://github.com/user-attachments/assets/8a6c3978-a076-4d0a-a051-6f470fffdba9" />

---

The above two screenshots are using the following set of icons:

<img width="820" alt="image" src="https://github.com/user-attachments/assets/23649b32-8d84-46f8-a1d6-bde2b808f577" />

---

<img width="918" alt="image" src="https://github.com/user-attachments/assets/db6f335a-7c1a-4a2d-84af-bb2dad376878" />

---

The above screenshot is using the following alternative set of icons:

<img width="879" alt="image" src="https://github.com/user-attachments/assets/e73b4180-d20f-4a64-a07e-5e7b6f62e955" />

---

But as I alluded to in the note at the top of the PR description: I don’t feel strongly about either of those sets of icons. Whatever other set might get decided on instead of either of those two would also be fine with me.

dlvhdr commented on README.md:
```diff
@@ -172,6 +172,7 @@ theme: # optional, see more info below
 pager:
   diff: less # or delta for example
 confirmQuit: false # show prompt on quit or not
+showAuthorRole: false # show author role ("new", "member", etc.)
```
Would this make more sense under `defaults.layout.prs`/`defaults.layout.issues` as well as an option to define it only for a specific section?
e.g.:
```yaml
prSections:
  - title: My OSS Project
    filters: is:open repo:dlvhdr/gh-dash
    layout:
      role:
        hidden: false
```

Not sure, would like to hear your opinion.

sideshowbarker replied:
The `showAuthorRole/ShowAuthorRole` config bits were left over from when I first started work on the patch and had thought it made sense for whether or not the author roles were shown to be configurable. But as I worked further on implementing the feature, I came to the conclusion that it seemed to make more sense to just always show the author roles — because doing it requires literally just one icon. So it’s unobtrusive, and I couldn’t imagine that most users would find it so useless or annoying as to want to not have it displayed.

But all that said, if you think it instead really should be made configurable — and I’ll update the patch to make it so.

dlvhdr replied:
Yeah U believe it should be configurable, as most repos O work against are work related, and I don't like the extra visual clutter. I do think it's a really addition for OSS maintainers. Gh-dash also supports defining sections for entirely different repos, some of which might not need these icons.

sideshowbarker replied:
Ok, thanks, I understand — I'll update the patch to make it configurable. The question then becomes: Should I make it on by default, or off be default?

dlvhdr replied:
Cool thank you! Hmm regarding default it's a coin toss from me. I guess turn it **on** by default.

dlvhdr commented on ui/constants/constants.go:
```diff
@@ -36,4 +36,25 @@ const (
+	// OwnerIcon          = "󱟜" // U+F17DC nf-md-crown_circle
+	// UnknownRoleIcon    = "󱐡" // U+F1421 nf-md-incognito_circle
+
+	NewContributorColor = "77"  // PaleGreen3
```
I suggest putting these under the `ui/theme/theme.go` file.
Maybe you can even reuse existing colors there, as those are configurable in the config file.

sideshowbarker replied:
OK — I’ve moved all the author-role icon colors to the `ui/theme/theme.go` file.

When I first started trying out the icon colors for this, I did actually use some existing colors. But then when I looked at in the built UI, it felt odd because it kind of seemed to unintentionally give the author-role icons the “meanings” from other existing icons and text in the UI; for example, purple meaning “merged”.

So I actually tried to come up with some new colors for the author-role icons that didn’t conflict in that way with the meanings from existing colored icons and text.

All that said, I’m really not a designer and I really don’t know much about trying to come up with color palettes that are complementary with each other etc. So I don’t feel strongly about what the colors should be, and if the ones I came up with in my patch are not the best choices — and if you want to re-use existing colors instead — then it’s 100% fine by me to change them in any way.

dlvhdr commented on config/parser.go:
```diff
@@ -313,6 +314,7 @@ func (parser ConfigParser) getDefaultConfig() Config {
 			},
 		},
 		ConfirmQuit: false,
+		ShowAuthorRole: false,
```
I'm not seeing where this is used? perhaps you intended to use it in `GetAuthor()`?

dlvhdr reviewed (commented):
This is really great! Thanks for the contrib! 🚀

dlvhdr commented:
forgot to mention, can you also add this to the docs website? Should be easy to find where to add it

sideshowbarker commented:
> forgot to mention, can you also add this to the docs website? Should be easy to find where to add it

I’ve pushed the other changes I made in response to review comments — but I’ve not yet made the updates to the docs site. I’ll make time to do that after ~12 hours or so from now.

sideshowbarker commented:
OK — I just now pushed an update (amended commit) with the following changes:

- moved both the definitions of the default color values for the icons, and the definitions of the icon strings/characters themselves into the `ui/theme/theme.go` and `config/parser.go` code
- added new config options for both the icon colors and the icons themselves
- added a new `showAuthorIcons` config option for flipping the author-icon display off
- updated the docs with details about all the new config options

https://github.com/dlvhdr/gh-dash/compare/175aa1a506ae7fa69b3b6a1dac301312409a3353..3efc99b7d0e55117b6687e053831a79ca022befd has the delta diff against what you’d previously reviewed.

dlvhdr commented on config/parser.go:
```diff
@@ -186,18 +209,20 @@ type UIThemeConfig struct {
 type ThemeConfig struct {
 	Ui     UIThemeConfig     `yaml:"ui,omitempty"     validate:"omitempty"`
 	Colors *ColorThemeConfig `yaml:"colors,omitempty" validate:"omitempty"`
+	Icons  *IconThemeConfig  `yaml:"icons,omitempty" validate:"omitempty"`
```
This is a great addition, thanks!

dlvhdr commented on ui/theme/theme.go:
```diff
@@ -8,31 +8,61 @@ import (
+	MemberIconColor         lipgloss.AdaptiveColor // config.Theme.Colors.Icon.Member
+	OwnerIconColor          lipgloss.AdaptiveColor // config.Theme.Colors.Icon.Owner
+	UnknownRoleIconColor    lipgloss.AdaptiveColor // config.Theme.Colors.Icon.UnknownRole
+	NewContributorIcon      string                 // config.Theme.Icons.NewContributor
```
Let's move the actual icons to `ui/constants/constants.go`

dlvhdr commented on ui/theme/theme.go:
```diff
@@ -8,31 +8,61 @@ import (
+	ContributorIcon:         "󱕾", // U+F157E nf-md-hand_heart_outline
+	CollaboratorIcon:        "", // U+EDA0  nf-fa-handshake_simple
+	MemberIcon:              "", // U+ED9F  nf-fa-handshake_angle
+	OwnerIcon:               "", // U+ED9F  nf-fa-handshake_angle
```
I think the `OwnerIcon` should more distinct than the collaborator and member icons. wdyt about "" (\uebb8), "" (\uedcf), "" (\uedd2), "󱇐" (\udb84\uddd0), "󰢏" (\udb84\udf60).

dlvhdr commented on config/parser.go:
```diff
@@ -186,18 +209,20 @@ type UIThemeConfig struct {
+	Theme           *ThemeConfig          `yaml:"theme,omitempty" validate:"omitempty"`
+	Pager           Pager                 `yaml:"pager"`
+	ConfirmQuit     bool                  `yaml:"confirmQuit"`
+	ShowAuthorIcons bool                  `yaml:"showAuthorIcons"`
```
I think this should also be defined in the section config.
This way, I can enable it globally but disable it for the "My PRs" section where it's irrelevant.

dlvhdr replied:
Something like:

```yaml
prSections:
  - title: Work
    filters: is:open repo:work/project
    layout:
      authorIcon:
        hidden: true
```

dlvhdr reviewed (commented):
Thanks for working on this!

sideshowbarker commented:
OK, thanks — I’ve just pushed an update (amended commit) that includes changes in response to all pending comments.

> I think the `OwnerIcon` should more distinct than the collaborator and member icons. wdyt about "" (\uebb8), "" (\uedcf), "" (\uedd2), "󱇐" (\udb84\uddd0), "󰢏" (\udb84\udf60).

Thanks — those look great. However, I feel a bit that users might potentially be confused somewhat by what the `\uedd2` “nf-fa-user_tie” icon is intended to convey. So, what about instead maybe the `\uedc6` “nf-fa-user_check” icon? Or if not that, another you can think of that might also be appropriate?

Anyway, what I have for now is this:

![image](https://github.com/user-attachments/assets/e765a992-486b-4ff7-9ad4-c628e69c9212)

Is that the correct mappings/order for what you had in mind? If not, please lemme know what I should change.

sideshowbarker commented:
https://github.com/dlvhdr/gh-dash/compare/3efc99b7d0e55117b6687e053831a79ca022befd..e1b03ca6ab6bf1772bf930b79f72f20a856a1616?w=1 is the delta against what you had previously reviewed

dlvhdr reviewed (commented):
This looks great, thanks!

dlvhdr reviewed (commented):
I think I would change the icons you chose, but for now it's fine.

dlvhdr commented:
I think that in a future PR I would like to add a label to the sidebar that explains what the icon means
